observability/src/test/java/io/grpc/observability/interceptors/InternalLoggingChannelInterceptorTest.java

```diff
@@ -0,0 +1,331 @@
+  private SettableFuture<Void> cancelCalled;
+  private SocketAddress peer;
+  private final Sink mockSink = mock(GcpLogSink.class);
+
+  @Before
+  public void setup() throws Exception {
+    factory = new InternalLoggingChannelInterceptor.FactoryImpl(mockSink);
+    interceptedListener = new AtomicReference<>();
+    actualClientInitial = new AtomicReference<>();
+    actualRequest = new AtomicReference<>();
+    halfCloseCalled = SettableFuture.create();
+    cancelCalled = SettableFuture.create();
+    peer = new InetSocketAddress(InetAddress.getByName("127.0.0.1"), 1234);
+  }
+
+  @Test
+  public void internalLoggingChannelInterceptor() throws Exception {
+    Channel channel = new Channel() {
+      @Override
+      public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
+          MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
+        return new NoopClientCall<RequestT, ResponseT>() {
+          @Override
+          @SuppressWarnings("unchecked")
+          public void start(Listener<ResponseT> responseListener, Metadata headers) {
+            interceptedListener.set((Listener<byte[]>) responseListener);
+            actualClientInitial.set(headers);
+          }
+
+          @Override
+          public void sendMessage(RequestT message) {
+            actualRequest.set(message);
+          }
+
+          @Override
+          public void cancel(String message, Throwable cause) {
+            cancelCalled.set(null);
+          }
+
+          @Override
+          public void halfClose() {
+            halfCloseCalled.set(null);
+          }
+
+          @Override
+          public Attributes getAttributes() {
+            return Attributes.newBuilder().set(Grpc.TRANSPORT_ATTR_REMOTE_ADDR, peer).build();
+          }
+        };
+      }
+
+      @Override
+      public String authority() {
+        return "the-authority";
+      }
+    };
+
+    @SuppressWarnings("unchecked")
+    ClientCall.Listener<byte[]> mockListener = mock(ClientCall.Listener.class);
+
+    MethodDescriptor<byte[], byte[]> method =
+        MethodDescriptor.<byte[], byte[]>newBuilder()
+            .setType(MethodType.UNKNOWN)
+            .setFullMethodName("service/method")
+            .setRequestMarshaller(BYTEARRAY_MARSHALLER)
+            .setResponseMarshaller(BYTEARRAY_MARSHALLER)
+            .build();
+
+    ClientCall<byte[], byte[]> interceptedLoggingCall =
+        factory.create()
+            .interceptCall(method,
+                CallOptions.DEFAULT,
+                channel);
+
+    // send client header
+    {
+      EventType expectedRequestHeaderEvent = EventType.GRPC_CALL_REQUEST_HEADER;
+      ArgumentCaptor<GrpcLogRecord> captor = ArgumentCaptor.forClass(GrpcLogRecord.class);
+      Metadata clientInitial = new Metadata();
+      interceptedLoggingCall.start(mockListener, clientInitial);
+      verify(mockSink).write(captor.capture());
+      assertEquals(captor.getValue().getEventType(),
```

Comment:
How about verifying other fields of `GrpcLogRecord` ? You can probably create a helper method to test these or at least some of these. It will be helpful to do this test with a non-empty client header

Reply:
Verifying other fields of `GrpcLogRecord` by using a non-empty client header. 
Added TODO to create a helper method for all events.